kernel/src/scan/state_info.rs

```diff
@@ -0,0 +1,582 @@
+    /// Get the state needed to process a scan.
+    ///
+    /// `logical_schema` - The logical schema of the scan output, which includes partition columns
+    /// `table_configuration` - The TableConfiguration for this table
+    /// `predicate` - Optional predicate to filter data during the scan
+    /// `classifier` - The classifier to use for different scan types. Use `()` if not needed
+    pub(crate) fn try_new<C: TransformFieldClassifier>(
+        logical_schema: SchemaRef,
+        table_configuration: &TableConfiguration,
+        predicate: Option<PredicateRef>,
+        classifier: C,
+    ) -> DeltaResult<Self> {
+        let partition_columns = table_configuration.metadata().partition_columns();
+        let column_mapping_mode = table_configuration.column_mapping_mode();
+        let mut read_fields = Vec::with_capacity(logical_schema.num_fields());
+        let mut metadata_field_names = HashSet::new();
+        let mut transform_spec = Vec::new();
+        let mut last_physical_field: Option<String> = None;
+
+        // We remember the name of the column that's selecting row indexes, if it's been requested
+        // explicitly. this is so we can reference this column and not re-add it as a requested
+        // column if we're _also_ requesting row-ids
+        let mut selected_row_index_col_name = None;
+
+        // This iteration runs in O(supported_number_of_metadata_columns) time since each metadata
+        // column can appear at most once in the schema
+        for metadata_column in logical_schema.metadata_columns() {
+            if let Some(MetadataColumnSpec::RowIndex) = metadata_column.get_metadata_column_spec() {
+                selected_row_index_col_name = Some(metadata_column.name().to_string());
+            }
+            metadata_field_names.insert(metadata_column.name());
+        }
+
+        // Loop over all selected fields and build both the physical schema and transform spec
+        for (index, logical_field) in logical_schema.fields().enumerate() {
+            if let Some(spec) =
+                classifier.classify_field(logical_field, index, &last_physical_field)
+            {
+                // Classifier has handled this field via a transformation, just push it and move on
+                transform_spec.push(spec);
+            } else if partition_columns.contains(logical_field.name()) {
+                // This is a partition column: needs to be injected via transform
+
+                // First ensure we don't have a metadata column with same name as a partition column
+                if logical_field.is_metadata_column() {
+                    return Err(Error::Schema(format!(
+                        "Metadata column names must not match partition columns: {}",
+                        logical_field.name()
+                    )));
+                }
+                // push the transform for this partition column
```

Comment:
This can be done as a check above in `for metadata_column in logical_schema.metadata_columns()`